feat/enhancements

epiqueras opened the pull request:
This PR hopefully finalizes the boilerplate standard for now.

* Make app more generic to Ethereum.
* Group common state (accounts, balance) into wallet.
* Add tests to 100% coverage. Closes #4 
* Add Identicon (react-blockies). Closes #5 

I also went around and standardized import/exports to follow a common standard (The index files in utils/components were causing Webpack to choke up when ran though jest or storybook because of circular dependencies):

* import regular named exports directly from their file using destructuring import syntax
e.g.
```js
import { name } from '.../utils/something'
```
* import actions, selectors, and shapes using the ```* as``` syntax
e.g.
```js
import * as somethingActions from '.../actions/something'

somethingActions.FETCH_SOMETHING
somethingActions.fetchSomething()
```
```js
import * as somethingSelectors from '.../reducers/something'

somethingSelectors.somethingShape.isRequired
somethingSelectors.getSomething(state)
```

epiqueras committed:
ci(linter): enforce component type checking through prop-types

epiqueras committed:
fix(styles): stop importing .scss files and import .css files

Add normalize.css.

epiqueras committed:
feat(components): make requires-meta-mask a reusable component

epiqueras committed:
refactor: normalize code style and make app more generic

epiqueras committed:
feat: handle fetch and create errors and abstract common logic

epiqueras committed:
feat(balance): add Identicon component using react-blockies

Closes #5

epiqueras committed:
feat(wallet): integrate account fetching logic into sagas

epiqueras committed:
feat(reducers): enhance UX drawing inspiration from Apollo Client

epiqueras committed:
test: one hundred percent coverage

Closes #4

satello commented on README.md:
```diff
@@ -15,14 +15,14 @@
-DEV_ARBITRATOR_ADDRESS=0xaea35f89f98996ae06aac344ab4b9ce1731059c4
+REACT_APP_DEV_ETHEREUM_PROVIDER=http://localhost:8545
+REACT_APP_DEV_STORE_PROVIDER=https://kleros.in
+REACT_APP_DEV_ARBITRATOR_ADDRESS=0xaea35f89f98996ae06aac344ab4b9ce1731059c4
```
This address has changed. I think you might just want to leave this blank in the boilerplate as it is likely to change a lot

satello commented on src/bootstrap/initializer.js:
```diff
@@ -1,30 +1,46 @@
+import * as walletSelectors from '../reducers/wallet'
+import { eth } from './kleros'
+import { renderIf } from '../utils/react-redux'
+import RequiresMetaMask from '../components/requires-meta-mask'
```
I like components being PascalCase

epiqueras replied:
Do you mean I should change the filename/path? Or is this good?

The only reason I made the filename kebab-case was so we can enforce a consistent filename pattern with the linter. Otherwise it would be up to us to make sure only component files have PascalCase filenames. I guess it's not that bad. What do you think?

satello commented on src/bootstrap/initializer.js:
```diff
@@ -1,30 +1,46 @@
+    const { isWeb3Loaded } = this.state
+    const { accounts, children } = this.props
+
+    return renderIf(
```
nice

satello commented on src/bootstrap/kleros.js:
```diff
@@ -1,21 +1,17 @@
 
-export default new Kleros(web3.currentProvider, process.env.STORE_PROVIDER)
+export { eth }
+export default new Kleros(eth.currentProvider, process.env.STORE_PROVIDER)
```
:+1:

satello commented on src/constants/errors.js:
```diff
@@ -1,2 +1,2 @@
-export const WEB3_NOT_RESOLVED =
-  'The object `web3` was not found. Maybe try to install Metamask https://chrome.google.com/webstore/detail/metamask/nkbihfbeogaeaoehlefnkodbefgpgknn?hl=en'
+export const ETH_NO_ACCOUNTS =
+  'web3 accounts were not found. Maybe try installing MetaMask: https://chrome.google.com/webstore/detail/metamask/nkbihfbeogaeaoehlefnkodbefgpgknn?hl=en'
```
I think the wording should be something like `web3 accounts not found. Do you have MetaMask installed? ...`

satello commented on src/containers/balance/index.js:
```diff
@@ -1,31 +1,71 @@
 import React, { PureComponent } from 'react'
```
what is this balance page for? is it just an example container?

epiqueras replied:
Yeah, I needed an example container so I could write an integration test with the whole redux saga fetch/loading/{receive,failed} flow.

satello reviewed (commented):
Nice work! My biggest overarching concern is there is quite a bit of code in here that currently wouldn't be used in the DApp such as the all the `eth balance` stuff and some of the utils and helpers. If we think we will never use it its probably worth it to take it out to reduce the clutter

epiqueras committed:
docs(readme): replace volatile env vars with placeholders

epiqueras committed:
feat(errors): change ETH_NO_ACCOUNTS message

epiqueras commented:
@satello 
I changed the readme .env file to use placeholders.

I changed the ETH_NO_ACCOUNTS message.

I got rid of all the utils that are not being used. Do you think there is anything else I should remove?

I also made the app more "generic" by changing `bootstrap/kleros.js` to `bootstrap/dapp-api.js`.

epiqueras committed:
refactor(utils): remove unused utils

epiqueras committed:
refactor(bootstrap): make app more generic with kleros='>dapp-api'

n1c01a5 commented on src/utils/create-reducer.js:
```diff
@@ -14,31 +23,28 @@ export default function createReducer(initialState, reducerMap) {
-      newState = { ...newState, ['loading' + resource]: false }
+    for (const actionTypePrefix of Object.keys(automaticActionPluginMap)) {
+      const l = actionTypePrefix.length
+      const a = action.type.slice(0, l)
```
the name `var` `l` or `a`  it's not very explicit

n1c01a5 commented:
@satello I disagree with the `ETH balance` component most of dapp are need this metric and I think it's good to have an concrete interaction with the blockchain

n1c01a5 commented:
@epiqueras good work
